components/ws-daemon/pkg/daemon/markunmount.go

```diff
@@ -101,8 +101,24 @@ func (c *MarkUnmountFallback) WorkspaceUpdated(ctx context.Context, ws *dispatch
 	}
 	ttl := time.Duration(gracePeriod)*time.Second + propagationGracePeriod
 
+	dispatch.GetDispatchWaitGroup(ctx).Add(1)
 	go func() {
-		time.Sleep(ttl)
+		defer dispatch.GetDispatchWaitGroup(ctx).Done()
+
+		defer func() {
+			// We expect the container to be gone now. Don't keep its referenec in memory.
+			c.mu.Lock()
+			delete(c.handled, ws.InstanceID)
+			c.mu.Unlock()
+		}()
+
+		wait := time.NewTicker(ttl)
+		defer wait.Stop()
+		select {
+		case <-ctx.Done():
+			return
+		case <-wait.C:
```

Comment:
> <-time.After(ttl)

Doesn't this leak a timer in case it does not fire? It's the vast minority of cases, granted, but the reason I preferred the explicit way (`NewTicker` + `defer ...Stop()`)

Reply:
ah only since Go 1.23 it doesn't leak, and daemon is on 1.22:

```
// As of Go 1.23, the garbage collector can recover unreferenced,
// unstopped timers. There is no reason to prefer NewTimer when After will do.
```

somewhere it was missing a defer though

Reply:
> ah only since Go 1.23 it doesn't leak

:bulb: TIL!

> somewhere it was missing a defer though

Will have a look :eyes:

Reply:
Found and fixed.